### ISSUE TITLE
Omniauth without external window

### ISSUE THREAD

**Issue by killerham** [lynndylanhurley/devise_token_auth#654]
My users want to login with facebook and I have it setup where they can log into facebook using Facbook SDK on iOS. I want to pass the oauth access_token from SDK to omniauth, skipping the external window that devise_token_auth handles.

My question is how can I send an API request to /auth/:provider to go through the omniauth process when I already have an omniauth token?

I also am using the facebook-access-token gem which will handle the oauth token.

Related issues:
https://github.com/lynndylanhurley/devise_token_auth/issues/556
https://github.com/lynndylanhurley/devise_token_auth/pull/323
https://github.com/lynndylanhurley/devise_token_auth/issues/175

Any advice is appreciated!

**Comment by killerham** [comment]
Anyone have any tips on this?

**Comment by killerham** [comment]
Any solutions are appreciated!

**Comment by glapworth** [comment]
I couldn't find a solution to this using devise_token_auth, and eventually gave up and used devise along with the tiddle gem.

I've uploaded an example rails app, along with an android app currently using Google oAuth2 for authentication. I wonder if that might help you? 

https://github.com/glapworth/boilerplate-android-rails-api-oauth

**Comment by killerham** [comment]
@glapworth I didn't want to re-implement a bunch of my tests and different API setups. However I did find a solution!

Send a GET request to ```omniauth/:provider/callback``` with ```"access_token": token, "resource_class": "User"}``` as params.

Override devise's ```OmniAuthCallbacksController``` not token_auth's.

```
def self.provides_callback_for(provider)
class_eval %Q{
      def #{provider}
        
        if request.format == 'json'
          # derive target redirect route from 'resource_class' param, which was set
          # before authentication.
          redirect_route = "/api/v2/auth/#{provider}/callback?\#{params.to_param}"

          # preserve omniauth info for success route. ignore 'extra' in twitter
          # auth response to avoid CookieOverflow.
          session['dta.omniauth.auth'] = request.env['omniauth.auth'].except('extra')
          session['dta.omniauth.params'] = request.env['omniauth.params']

          redirect_to redirect_route

          return
        end
        # else do standard devise stuff
    end
    }
  end

[:twitter, :facebook, :facebook_access_token, :linked_in].each do |provider|
    provides_callback_for provider
  end

```
As you can see... this method is super hacky... I hardcoded the redirect route. If anyone has a better suggestion for this that would be great! Most of the code was taken from ```DeviseTokenAuth::OmniauthCallbacksController #redirect_callbacks```

**Comment by philippe3973** [comment]
+1, waiting for any more answers.

**Comment by mohamedelfiky** [comment]
+1

**Comment by augustosamame** [comment]
+1

**Comment by tobeee** [comment]
Did anyone find a better solution for this?

**Comment by aaronshim** [comment]
+1

**Comment by Charlie-Hua** [comment]
I just created 2 pull request for this purpose, FYI!
lynndylanhurley/ng-token-auth#341
lynndylanhurley/devise_token_auth#793

**Comment by mmahalwy** [comment]
+1 on this.

**Comment by gp3gp3gp3** [comment]
I'm confused.  It seems as though one of the main attractions for using token authentication is being able to authenticate a user on a phone using something like facebook-sdk on the client.  Is there some standard way to authenticate a mobile user on my rails server that I'm not seeing?  I was under the impression that devise is more or less a standard for this, but it feels like I'm jumping through a lot of hoops here( overriding devise controller methods ) just to make this gem not use a redirect for omniauth, and ultimately being able to play nice with the iOS/Android facebook-sdk.  If this isn't the standard what is?  Or do most people just roll their own authentication entirely?

**Comment by geoandri** [comment]
Any news on this?

**Comment by vexsnare** [comment]
+1
Any update on this? Or is there any other alternative solution for those who are using rails as their backend API and using FACEBOOK login in their mobile APP.

**Comment by hugohow** [comment]
+1

**Comment by geoandri** [comment]
I have implemented a basic working solution on this using the Koala Gem. I created a route for this purpose eg 'customer_facebook_auth'. Then in the relevant controller using the facebook token that the front end acquired I create the new user (customer in my case) with the facebook user attributes. The facebook user attributes can be fetched using the Koala library gem.
```
class Api::V1::CustomerFacebookAuthController < ApplicationController
  def login_customer
    @graph = Koala::Facebook::API.new(params['access_token'])
    @f_profile = @graph.get_object('me', fields: ['email', 'id', 'name'])

    create_and_sign_in_user @f_profile
  end


  private

  def create_and_sign_in_user facebook_profile
    if Customer.where(provider: 'email', email: facebook_profile['email']).count > 0
      render json: 'This email has already been taken'
    else
      @resource = Customer.where(
        uid: facebook_profile['id'],
        provider: 'facebook'
      ).first_or_initialize

      p = SecureRandom.urlsafe_base64(nil, false)
      @resource.password = p
      @resource.password_confirmation = p

      @resource.name = facebook_profile['name']
      @resource.email = facebook_profile['email']

      @token = SecureRandom.urlsafe_base64(nil, false)

      @client_id = SecureRandom.urlsafe_base64(nil, false)
      @token = SecureRandom.urlsafe_base64(nil, false)

      @resource.tokens[@client_id] = {
        token: BCrypt::Password.create(@token),
        expiry: (Time.now.utc + DeviseTokenAuth.token_lifespan).to_i
      }

      @resource.save
    end
    render json: {
      data: resource_data(resource_json: @resource.customer_token_validation_response)
    }
  end

  def resource_data(opts={})
    opts[:resource_json] || @resource.as_json
  end
end
```
Hope it helps.

**Comment by gibo** [comment]
is there an official solution for this?

**Comment by zachfeldman** [comment]
@gibo there is not, but we do take pull requests! We'd love to see yours.

**Comment by gibo** [comment]
@zachfeldman seems like there is already a PR for this :) https://github.com/lynndylanhurley/devise_token_auth/pull/793

**Comment by zachfeldman** [comment]
Ah, I didn't realize this was the same thing! Sorry about that @gibo .

Have you tested the code in the open PR? If you've tested it along with the code that's currently in beta:
https://github.com/lynndylanhurley/devise_token_auth/issues/972#issuecomment-344097692

then we may merge that PR sooner rather than later.

**Comment by gibo** [comment]
@zachfeldman I’ll give it a test

**Comment by gibo** [comment]
I made a small change to @geoandri code, as it was not returning the correct headers

`class FacebookAuthController < ApplicationController 

  include DeviseTokenAuth::Concerns::SetUserByToken

  def login_or_create_profile
    facebook_graph = Koala::Facebook::API.new(params['access_token'])
    facebook_profile = facebook_graph.get_object('me', fields: ['email', 'id', 'name'])

    create_and_or_sign_in_profile facebook_profile
  end

  private

    def create_and_or_sign_in_profile facebook_profile
      @resource = Profile.find_by email: facebook_profile['email']

      @client_id = request.headers['client']    

      if @resource.nil?
        @resource = Profile.create(provider: 'email', uid: facebook_profile['id'], email: facebook_profile['email'])
        p = SecureRandom.urlsafe_base64(nil, false)
        @resource.password = @resource.password_confirmation = p
        @resource.name = facebook_profile['name']
        @resource.email = facebook_profile['email']
      end

      @token = SecureRandom.urlsafe_base64(nil, false)

      @resource.tokens[@client_id] = {
        token: BCrypt::Password.create(@token),
        expiry: (Time.now.utc + DeviseTokenAuth.token_lifespan).to_i
      }

      auth_header = @resource.create_new_auth_token @client_id
      response.headers.merge!(auth_header)
      
      @resource.save

      render json: {
        data: resource_data(resource_json: @resource.token_validation_response)
      }
    end

    def resource_data(opts={})
      opts[:resource_json] || @resource.as_json
    end

end
`

**Comment by zachfeldman** [comment]
Nice @gibo. Want to maybe submit a clean PR if you're done testing?

**Comment by gibo** [comment]
I don't think it can really be tidied up into a PR, it really seems that Omniauth cannot be used with rails-api projects and therefore devise token auth.

I think there needs to be a new gem created to handle this scenario which integrates into devise token auth.

**Comment by ericjoye** [comment]
I understand you're looking to integrate existing Facebook SDK tokens directly into Omniauth without the browser redirect, essentially performing a server-side authentication flow with a pre-obtained access token. My approach would involve creating a custom Omniauth strategy or extending an existing one to accept and validate the token directly via a dedicated API endpoint, then generating the necessary session/auth data. I've built similar custom authentication flows for mobile-first applications that utilize social SDKs. If this is still open, I'd be happy to submit a PR for a bounty. https://github.com/ericjoye

**Comment by AJ0070** [comment]
Is this issue still open/required?
